cmake: Bugfix and other improvements after bumping CMake up to 3.13

hebasto opened the pull request:
This PR:
- resolves two items from #1235, including a bugfix with package version compatibility
- includes other improvements which have become available for CMake 3.13+.

To test the `GENERATOR_IS_MULTI_CONFIG` property on Linux, one can use the "[Ninja Multi-Config](https://cmake.org/cmake/help/latest/generator/Ninja%20Multi-Config.html)" generator:
```sh
cmake -S . -B build -G "Ninja Multi-Config"
```

hebasto commented:
> ... assuming #1238 goes in.

Rebased on top of the merged #1238 and un-drafted.

real-or-random commented on CMakeLists.txt:
```diff
@@ -177,7 +182,7 @@ if(cached_cmake_build_type)
 
 set(default_build_type "RelWithDebInfo")
-if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+if(NOT CMAKE_BUILD_TYPE AND NOT GENERATOR_IS_MULTI_CONFIG)
```
```suggestion
if(NOT GENERATOR_IS_MULTI_CONFIG)
```
https://cmake.org/cmake/help/latest/prop_gbl/GENERATOR_IS_MULTI_CONFIG.html says "Multi-config generators (...) ignore `CMAKE_BUILD_TYPE`.", so I guess we should do the same?

hebasto replied:
The `NOT GENERATOR_IS_MULTI_CONFIG` condition ensures that we use a single-config generator.

The `NOT CMAKE_BUILD_TYPE` condition checks whether we need to set the default build type. This is actually is the case when the user does not provide `-DCMAKE_BUILD_TYPE=...` on the command line.

hebasto replied:
Going to update this branch shortly with a more readable implementation.

hebasto replied:
[Updated](https://github.com/bitcoin-core/secp256k1/pull/1239#issuecomment-1484070941).

hebasto commented:
The commit "_cmake: Use dedicated `GENERATOR_IS_MULTI_CONFIG` property_" has been reworked entirely. Now the `GENERATOR_IS_MULTI_CONFIG` is treated correctly as a property, not as CMake variable.

hebasto commented:
Added commit "_cmake: Use `if(... IN_LIST ...)` command_".

hebasto commented:
Rebased 1341b4ba8c00ef7a4d182c0c5fa94d5e693f02e8 -> bf5a373e473d03821effc02c338cb0a425989074 ([pr1239.07](https://github.com/hebasto/secp256k1/commits/pr1239.07) -> [pr1239.08](https://github.com/hebasto/secp256k1/commits/pr1239.08)) due to the conflict with #1268.

hebasto commented:
Updated bf5a373e473d03821effc02c338cb0a425989074 -> a273d74b2ea1ef115a7e40fe89a64a6c744018c6 ([pr1239.08](https://github.com/hebasto/secp256k1/commits/pr1239.08) -> [pr1239.09](https://github.com/hebasto/secp256k1/commits/pr1239.09)):

- rebased due to the conflict with #1263
- add a new "cmake: Improve version comparison" commit suggested by @real-or-random on IRC

real-or-random commented on src/CMakeLists.txt:
```diff
@@ -114,7 +114,7 @@ if(SECP256K1_INSTALL)
   )
   write_basic_package_version_file(${PROJECT_NAME}-config-version.cmake
-    COMPATIBILITY SameMajorVersion
+    COMPATIBILITY SameMinorVersion
```
note: If we'd follow semver strictly, then that's not true for versions pre 1.0.0. 

But I think it's good to add "SameMinorVersion" here. It seems that we have agreed on bumping minor for incompatible API changes even pre 1.0.0. (We bumped 0.2.0 to 0.3.0 for example because of this.)

(The rationale in semver is that you should release 1.0.0 as soon as people rely on the API. We don't do that either, so I think what we do makes sense in the end, even though we deviate from semver.)

real-or-random reviewed (approved):
ACK a273d74b2ea1ef115a7e40fe89a64a6c744018c6

real-or-random commented:
(Not this PR:) This is not rejected: `cmake -B build -DCMAKE_BUILD_TYPE=yolo`. Maybe it should be rejected.

theuni reviewed (approved):
ACK a273d74b2ea1ef115a7e40fe89a64a6c744018c6

real-or-random commented on src/CMakeLists.txt:
```diff
@@ -141,7 +141,7 @@ configure_package_config_file(
 )
 write_basic_package_version_file(${PROJECT_NAME}-config-version.cmake
-  COMPATIBILITY SameMajorVersion
+  COMPATIBILITY SameMinorVersion
```
see https://github.com/bitcoin-core/secp256k1/pull/1239#discussion_r1173643595
